### PR TITLE
fix(ekf2): preserve EV position bias on reset when GNSS inactive

### PR DESCRIPTION
**Pull request by dakejahl**
Supersedes https://github.com/PX4/PX4-Autopilot/pull/21683

## Problem

When fusing GNSS + EV position, losing GNSS triggers an EV position reset that discards the learned bias (`_ev_pos_b_est.reset()`). This snaps the local position to the raw EV measurement, which may have drifted significantly from the current estimate. In HOLD mode this causes the drone to fly toward what it now thinks is the correct waypoint position.

## Fix

Preserve the estimated EV bias through the reset by using `measurement - _ev_pos_b_est.getBias()` instead of resetting to the raw measurement. This is consistent with what the code already does in the `gnss_pos` branch of the fusion-failing recovery path (line 286).

Fixed in two places in `updateEvPosFusion()`:
- The EV reset path (triggered by `ev_reset`, yaw alignment change, or bias estimator state change)
- The fusion-failing recovery path (when `_nb_ev_pos_reset_available > 0`)

## Broader scope / future work

There are related issues in the GNSS + EV interaction that go beyond this fix:

- **GNSS re-acquisition after dropout**: @xdwgood reported on #21683 that while this fix helps when GNSS drops, re-acquiring GNSS afterwards can cause a large position jump. The reverse transition (EV-only → GNSS+EV) needs similar bias-aware handling.
- **`startEvPosFusion` without GNSS**: When EV fusion starts without GNSS (line 225-226), the bias is also unconditionally reset. If a meaningful bias was learned from a prior GNSS+EV session this could cause a similar jump, though on first start the bias is zero so this is lower priority.
- **Explicit GPS local frame flag**: @dagar noted the need for a `control_status` flag to explicitly track whether we're in the GPS local frame (origin initialized, yaw aligned north, etc.), rather than inferring it from `gnss_pos`.
- **EV yaw offset on GNSS loss**: When GNSS drops and we continue with EV, should we allow `ev_yaw` with the yaw offset captured at switch time? Currently EV yaw is blocked when `gps` flag is set.
- **Full frame reset on EV loss**: If `ev_pos`/`ev_vel` is lost or resets while operating without GNSS, should we fully reset (bias, origin, yaw_align) and operate entirely in the vision frame?
- **SITL testing**: A fake_ev module with configurable heading offset and origin would make it easier to test these edge cases.

## Test plan

- [ ] SITL: fly a mission with GNSS + EV, disable GNSS mid-flight, verify no position jump
- [ ] SITL: verify EV fusion-failing recovery path also preserves position
- [ ] Review that bias is still correctly re-initialized when GNSS comes back

### CODE REVIEW

**Comment by dakejahl** [comment]
@henrykotze you also may be interested in this

**Comment by henrykotze** [comment]
@dakejahl will see if i can get to this in the next few weeks with SITL testing.

File: src/modules/ekf2/EKF/aid_sources/external_vision/ev_pos_control.cpp [hunk 1]
```diff
@@ -287,8 +286,7 @@ void Ekf::updateEvPosFusion(const Vector2f &measurement, const Vector2f &measure
 					_ev_pos_b_est.setBias(-getLocalHorizontalPosition() + measurement);
 
 				} else {
-					resetHorizontalPositionTo(measurement, measurement_var);
-					_ev_pos_b_est.reset();
+					resetHorizontalPositionTo(measurement - _ev_pos_b_est.getBias(), measurement_var);
```
Comment (henrykotze) [inline src/modules/ekf2/EKF/aid_sources/external_vision/ev_pos_control.cpp]: Since we are making use of the bias when resetting the state, shouldn't we do the same when resetting the variance of the state? 

ie:
```resetHorizontalPositionTo(measurement - _ev_pos_b_est.getBias(), measurement_var + _ev_pos_b_est.getBiasVar());```

**Comment by henrykotze** [comment]
Initial testing is working nicely. There is no transient when the state get reset to the ev position measurement when the bias is taken into account. 
image

Here is the same simulation, without the bias taken into account:
image


*Take note*: 
I only removed the bias in the xy axis in the second picture attached, and thus there is no transient in the z-axis.

**Comment by dakejahl** [comment]
@henrykotze would you mind taking over this branch and making the requested changes? If it looks good to you I will merge

**Comment by henrykotze** [comment]
@dakejahl, will do :+1: 
Will attempt to see if i can get the GNSS re-acquisition implemented as well

**Comment by dakejahl** [comment]
@henrykotze Thanks! If you do that, maybe do it in a separate stacked PR that will follow-on right after this one so we can keep the conversation history linear.

**Comment by dakejahl** [comment]
Claude found one more additional issue, see latest commit https://github.com/PX4/PX4-Autopilot/pull/26794/changes/084f33e81d3671a8e486dba433fa1612c628bffe

**Comment by henrykotze** [comment]
I pushed to my fork: https://github.com/henrykotze/PX4-Autopilot/tree/fix/ekf2-ev-preserve-bias

* EV is now able use the learned bias after failure when it starts again 
* No jump in position estimate when GPS fails (includes  the bias when resetting the position estimates)
* EV also resets the bias when GPS gets regained, however with extended flights there will probably be a position estimate jump in longer flights where more drift can occur.
* In the log below, I used quite large offsets( 2m and 3m position offset in my EV measurements) 

image

Pretty happy with this, just need to do another round of checks. 

@dakejahl , do you want me to make new PR and reference this one? ( happy to do so) As i cannot push to this branch. 
I also added a small change to PositionBiasEstimator, to allow us to check whether the bias was fused when a aid source fails and then start fusing it again. 

But I will make it all clearer when im happy with my code :+1:

**Comment by dakejahl** [comment]
@henrykotze check your email, you can push to this branch directly

**Comment by henrykotze** [comment]
@dakejahl LGTM

Just a few comments for when people refer back to this PR:

If the EV has learned a bias, it will again make use of it during reinitialize after failure when VIO recovers after a period where both GNSS and VIO were unavailable.

There are pro and cons of doing this

* The bias may no longer be geometrically valid after a long outage, introducing an unnecessary offset into the position estimate (mitigated by bias reset on GNSS recovery)
* The bias covariance grows during the outage, meaning the EKF's confidence in the stored bias degrades the longer VIO was unavailable, causing the measurement variance to be higher. 

Pros:

* For short VIO outages without GNSS, the previously learned bias remains geometrically valid and prevents a position jump on VIO recovery
* The EKF naturally self-corrects the stale bias risk — covariance growth during the outage causes the filter to weight the reused bias less aggressively the longer the outage was
* The bias is explicitly reset when GNSS is regained, bounding the worst-case impact of an invalid bias
* Avoids a cold reinitialisation of the bias on every VIO recovery, which would cause a transient position error in the short-outage case

File: src/modules/ekf2/EKF/bias_estimator/position_bias_estimator.hpp [hunk 2]
```diff
@@ -50,8 +50,9 @@ class PositionBiasEstimator
 	virtual ~PositionBiasEstimator() = default;
 
 	bool fusionActive() const { return _is_sensor_fusion_active; }
+	bool wasFusionActive() const { return _was_fusion_active; }
```
Comment (mrpollo) [inline src/modules/ekf2/EKF/bias_estimator/position_bias_estimator.hpp:56]:
```suggestion
	// Latched true the first time bias fusion activates (GNSS+EV co-fusing), and
	// intentionally never cleared. Used on EV restart without GNSS to decide whether
	// a previously-learned bias exists and can be reused, rather than cold-resetting.
	bool wasFusionActive() const { return _was_fusion_active; }
```

**Review by mrpollo — DISMISSED** [review]

**Comment by mrpollo** [comment]
I'm merging as soon as CI is green

**Review by mrpollo — APPROVED** [review]
